evm/src/fuzz/invariant/executor.rs

```diff
@@ -0,0 +1,420 @@
+pub struct InvariantExecutor<'a> {
+    pub executor: &'a mut Executor,
+    /// Proptest runner.
+    runner: TestRunner,
+    /// Contracts deployed with `setUp()`
+    setup_contracts: &'a BTreeMap<Address, (String, Abi)>,
+    /// Contracts that are part of the project but have not been deployed yet. We need the bytecode
+    /// to identify them from the stateset changes.
+    project_contracts: &'a BTreeMap<ArtifactId, (Abi, Vec<u8>)>,
+}
+
+impl<'a> InvariantExecutor<'a> {
+    /// Instantiates a fuzzed executor EVM given a testrunner
+    pub fn new(
+        executor: &'a mut Executor,
+        runner: TestRunner,
+        setup_contracts: &'a BTreeMap<Address, (String, Abi)>,
+        project_contracts: &'a BTreeMap<ArtifactId, (Abi, Vec<u8>)>,
+    ) -> Self {
+        Self { executor, runner, setup_contracts, project_contracts }
+    }
+
+    /// Fuzzes any deployed contract and checks any broken invariant at `invariant_address`
+    /// Returns a list of all the consumed gas and calldata of every invariant fuzz case
+    pub fn invariant_fuzz(
+        &mut self,
+        invariant_contract: InvariantContract,
+        test_options: InvariantTestOptions,
+    ) -> eyre::Result<Option<InvariantFuzzTestResult>> {
+        let (fuzz_state, targeted_contracts, strat) =
+            self.prepare_fuzzing(&invariant_contract, test_options)?;
+
+        // Stores the consumed gas and calldata of every successful fuzz call.
+        let fuzz_cases: RefCell<Vec<FuzzedCases>> = RefCell::new(Default::default());
+
+        // Stores data related to reverts or failed assertions of the test.
+        let failures =
+            RefCell::new(InvariantFailures::new(&invariant_contract.invariant_functions));
+
+        let blank_executor = RefCell::new(&mut *self.executor);
+
+        // Make sure invariants are sound even before starting to fuzz
+        if assert_invariants(
+            &invariant_contract,
+            &blank_executor.borrow(),
+            &[],
+            &mut failures.borrow_mut(),
+        )
+        .is_err()
+        {
+            fuzz_cases.borrow_mut().push(FuzzedCases::new(vec![]));
+        }
+
+        if failures.borrow().broken_invariants_count < invariant_contract.invariant_functions.len()
+        {
+            // The strategy only comes with the first `input`. We fill the rest of the `inputs`
+            // until the desired `depth` so we can use the evolving fuzz dictionary
+            // during the run. We need another proptest runner to query for random
+            // values.
+            let branch_runner = RefCell::new(self.runner.clone());
+            let _ = self.runner.run(&strat, |mut inputs| {
+                // Scenarios where we want to fail as soon as possible.
+                {
+                    if test_options.fail_on_revert && failures.borrow().reverts == 1 {
+                        return Err(TestCaseError::fail("Revert occurred."))
```

Comment:
Is there any way to return the actual revert message that occured?

Reply:
The moment we encounter it, we actually store the reason on `can_continue(...)`.

I just haven't found a way to quickly exit the `run` closure...